cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache.go

```diff
@@ -122,35 +124,37 @@ func newCacheWithRedisClient(config cacheConfig, redisCli *redis.Client) (*Resou
 	log.Infof("[PING] -> [%s]", pong)
 
 	c := ResourceWatcherCache{
-		config:         config,
-		watcherStarted: false,
-		watcherMutex:   sync.Mutex{},
-		redisCli:       redisCli,
+		config:    config,
+		initOk:    false,
+		initMutex: sync.Mutex{},
+		redisCli:  redisCli,
 	}
+
+	c.initMutex.Lock()
 	go c.startResourceWatcher()
+
 	return &c, nil
 }
 
 func (c *ResourceWatcherCache) startResourceWatcher() {
 	log.Infof("+ResourceWatcherCache startResourceWatcher")
-	c.watcherMutex.Lock()
```

Comment:
> Why did you move the mutex lock from here into the `newCacheWithRedisClient` above, rather than leaving both the lock and unlocks all in the one method here? It always seems easier to reason about if they're both in the one method IMO, but perhaps there's something I'm missing?

There is a race condition as startResourceWatcher is launched in a go routine. I wanted to make sure that I flag the fact that we are initializing before launching a concurrent thread

Reply:
> More generally, it's not clear to me why you're using a mutex here? It's used in two places that I can see:
> 
>     1. When creating a resource watcher (which also starts it), even though there's only ever a single go routine watching resources (ie. no concurrency) per watcher, you use it to guard setting `initOk`, and
> 
>     2. When a watcher is processing an event, you read `initOk` during `checkInit()`
> 
> 
> So again, I am probably missing something, but I don't seed the need for the mutex? In fact, I would have thought it'd be much simpler to have `newRedisCacheClient` only return without an error if it _had_ started without an error (ie. not even need an `initOk` member. The only thing that returns an error in `startResourceWatcher()` is `newResourceWatchChan`, so why not just create the watcher channel in `newRedisCacheClient` (and error appropriately), then just do `go c.processEvents()` straight after (still in `newRedisCacheClient` before returning without error - that way, you only ever get a new redis cache client if it _is_ initialized (and there's no longer any need for "we should never reach here" type comments)?
> 
> Apologies in advance if I've completely missed an obvious reason why that's not simpler or not possible.

I see your point, let me see if I can re-write it the way you suggest

Reply:
well, there appears to be some weird behavior I can't explain yet that's blocking me from making this change as you described. If I create a channel in one thread and then wait for events on that channel in another thread, the k8s fake watcher is broken - the events are never delivered. So the unit tests hang. I am looking into why that is and what the solution may be

Reply:
the latest: while trying to figure out why the watcher wasn't receiving events I came accross a piece of code on the web that gracefully handles the case when k8s happens to lose (close) the channel for whatever reason. I had briefly mentioned this in https://github.com/kubeapps/kubeapps/issues/3032#issuecomment-876033943 but forgot to do anything about. Anyway, long story short, the watch/channel may die and need to be re-started at any point. Given that, I think it made sense to create the channel inside startResourceWatcher (and re-create it there if needed). Given that, I think I need a mutex in order to make sure my callers don't try to use the cache while its in half-initialized state

Reply:
Makes sense, but if that's the case, you'd need to ensure you don't re-get the older events? Actually, I'd expect something should be available to handle that for you? Checking, it looks like what you're after is https://pkg.go.dev/k8s.io/client-go@v0.21.0/tools/watch#RetryWatcher . If you use that in your `newResourceWatcherChan()` to wrap your current watcher, it should be handled for you (including ensuring you don't get the older messages etc.), including ensuring that from the client's pov, "it will get restarted from the last point without the consumer even knowing about it". So the channel from the retry watcher should *not* be closed when the watcher stops, so we'd not need to handle any recreation or `initOk` state etc.?

Reply:
I think RetryWatcher is meant to work with conjunction with https://pkg.go.dev/k8s.io/client-go@v0.21.3/tools/cache which isn;t useful to us because we are not using client-side caching. We use redis (server-side) caching.

Reply:
Yep, up to you. I'll leave a +1 so you can do what you think best.

Reply:
For what its worth, I have been looking at RetryWatcher for some time now. It looks like all it relies on is cache.Watcher, which is not as scary as I thought (just one interface method "Watch") and having object resourceVersion. In other words, not necessarily as dependent on the whole client-side cache machinery as I originally thought. So I am going to try and make it work over the weekend. To be continued...

P.S. I love (good) movies. So, a quote of one of my favorites (I'll let you guess which one): 
 "What is the most resilient parasite? Bacteria? A virus? An intestinal worm? An idea. Resilient... highly contagious. Once an idea has taken hold of the brain it's almost impossible to eradicate. An idea that is fully formed - fully understood - that sticks; right in there somewhere"

Reply:
Update: I am still working on it. There are a couple of things that I would like to point out.

The current caching code in flux plug-in is broken. As one example why, imagine a state where 
the cache is in-sync with k8s store and both have repos A and B.
Now kubeapps-api server or the watcher goes down. The user then deletes the repo B via 
kubectl or some other means. Now kubeapps-apis server or the watcher comes back up
Redis DB still has A and B cached. The current code, which isn't setting resourceVersion
 in ListOptions when doing a Watch() call, will get a "synthetic" Add event for every 
 repo in kubeapps (see https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions under Watch:
resourceVersion unset). Side note: this behavior only really makes sense for when the cache starts out empty and there are already 
resources in k8s, but that's not the point here. The point is there is an "Add" event for A, which is harmless because it'll just cause a re-index of A and set the value to what it was earlier.
So, at the end, A and B will exist in redis cache and that IS a problem as B should no longer exist.

One possible solution: clear the cache (use something like https://redis.io/commands/FLUSHALL to do it) before starting a watch.

Now, this does not address the problem with watcher being robust and recovering from all errors. 
RetryWatcher MAY look like it will help but as it turns out there are a couple of issues with it as well:

1) it requires one to specify initialResourceVersion, so in practice that means I have to do a List() call first, flush and manually re-populate the cache with the results, get the resourceVersion from the said List call and  then use it to start the RetryWatcher. Kind of a hassle, but not impossible.
2) the bigger problem, IMO, is that RetryWatcher does not handle this 
https://kubernetes.io/docs/reference/using-api/api-concepts/#410-gone-responses, which I understand 
now to be a fairly common scenario. A given Kubernetes server will only preserve a historical list of 
changes for a limited time. RetryWatcher just stops watching when it encounters 410 as it considers 
this to be a non-retryable scenario.

Still working on the best solution. https://kubernetes.io/docs/reference/using-api/api-concepts tells me to use https://github.com/kubernetes/client-go/blob/v0.21.3/tools/cache/reflector.go but that looks way too complex  for what we need.

Reply:
I have coded a solution which uses RetryWatcher and I believe is robust. Working on unit tests now which is proving to be a bit a challenge. resourceVersion field is now critical part of the solution but the fake dynamic client is constraining me in what I can do. For example, 
`	list := unstructured.UnstructuredList{
		Object: map[string]interface{}{
			"metadata": map[string]interface{}{
				"resourceVersion": "1",
			},
		},
		Items: []unstructured.Unstructured{},
	}
	log.Infof("1=======> [%s], %s", list.GetResourceVersion(), prettyPrintMap(list.Object))

	dynamicClient := fake.NewSimpleDynamicClientWithCustomListKinds(
		runtime.NewScheme(),
		map[schema.GroupVersionResource]string{
			{Group: fluxGroup, Version: fluxVersion, Resource: fluxHelmRepositories}: fluxHelmRepositoryList,
		},
		&list)

	list2, err := dynamicClient.Resource(schema.GroupVersionResource{Group: fluxGroup, Version: fluxVersion, Resource: fluxHelmRepositories}).Namespace("").List(context.Background(), v1.ListOptions{})
	if err != nil {
		return nil, nil, nil, err
	}

	log.Infof("2=======> [%s], %s", list2.GetResourceVersion(), prettyPrintMap(list2.Object))
`  
logs 
`I0717 22:36:38.944870    3428 server_test.go:1479] 1=======> [1], {
  "metadata": {
    "resourceVersion": "1"
  }
}
I0717 22:36:38.945478    3428 server_test.go:1493] 2=======> [], {
  "apiVersion": "source.toolkit.fluxcd.io/v1beta1",
  "kind": "HelmRepositoryList",
  "metadata": {
    "resourceVersion": ""
  }
}
`
i.e. we've lost the list resourceVersion somewhere in the process, which doesn't happen on k8s, only with fake client. Still working on it...

Reply:
> For what its worth, I have been looking at RetryWatcher for some time now. It looks like all it relies on is cache.Watcher, which is not as scary as I thought (just one interface method "Watch") and having object resourceVersion. In other words, not necessarily as dependent on the whole client-side cache machinery as I originally thought.

Great.

> So I am going to try and make it work over the weekend.

Not great, but I can't stop you. JFYI, with some exceptions, I'll generally be spending time with the other priorities around me outside of work on the weekends :)

> To be continued...
> 
> P.S. I love (good) movies. So, a quote of one of my favorites (I'll let you guess which one):
> "What is the most resilient parasite? Bacteria? A virus? An intestinal worm? An idea. Resilient... highly contagious. Once an idea has taken hold of the brain it's almost impossible to eradicate. An idea that is fully formed - fully understood - that sticks; right in there somewhere"

I was actually looking at that movie (not watching) the other day as an option to watch with my older kids. And yes, it works with both good and bad ideas. Hard to protect against sometimes too.

Anyway, just let me know if want me to look at something specific in your PR, or if you want me to go over it once your tests are done.

Reply:
yeah, here is to hoping it's not a bad one :-)
I think I will ask for a review, once I am done, as I've made some significant rewrites

Reply:
Yeah, I think it'd be best if we always review non-trivial changes (ie. chart bumps or similar) but I'll hold back and just look at essentials for yours :)

Reply:
Yeah, this latest commit should have most of the changes. The most important part is in cache.go. That's what I've been trying to make work for the last several days. Basically, what happens when things "go bad" (kubeapps-apis pod or watcher dies for whatever reason and needs to restart). I have done some initial manual testing and all appears to work as expected, i.e. the cache stays in sync with whats in k8s. I feel somewhat confident about the resiliency of the code (I didn't before). If you would like to review the logic there, I'd certainly welcome the feedback. If not, that's fine too. I will probably spend another day writing a unit test or two before merge. They turn out to take the bulk of the work (all this "faking" of this and that, some of the stuff I've had to come up to the unit test to work looks a bit gross). I am not complaining :-) Cheers

Reply:
> Yeah, this latest commit should have most of the changes. The most important part is in cache.go. [...]  If you would like to review the logic there, I'd certainly welcome the feedback.

Great, just took a look at that commit and added another +1.

Reply:
thanks Michael